### ISSUE TITLE
Place Tree Underwater Tooltip Bug

### ISSUE THREAD

**Issue by RugnirViking** [OpenRCT2/OpenRCT2#8468]
**OS:** Windows 10
**Version:** 0.2.1-448
**Commit/Build:** ce3711d on dev branch

When attempting to place a tree on a lakebed, an incorrect tooltip is shown. It informs the player that the tree can only be placed on level ground, even if the tree is in fact able to be placed on sloped ground. This behaviour is different to vanilla, where the tooltip makes a lot more sense.

- [ NO ] Reproducible in RCT2 (vanilla)?
- [ NO ] Specific to multiplayer?

**Steps to reproduce:**
1. Open a park with water in
2. Try to place a tree that can be placed on sloped ground on a tile of water where the lake bed is flat. The example I used was the incense cedar tree.
3. Note that the tooltip reads "Can't position this here: Level land required"
4. In vanilla rct2, the tooltip is "Can't position this here: Can only build this on land!"

**Dump file**

**Screenshots / Video:**

**Save game:**

**Comment by ZehMatt** [comment]
Sorry I'm not entirely sure I understand the issue. Is only the text on the tooltip different or does the placement of the tree behave differently?

**Comment by RugnirViking** [comment]
> Sorry I'm not entirely sure I understand the issue. Is only the text on the tooltip different or does the placement of the tree behave differently?

It is just the text of the tooltip; the tree fails to place in both places, the tooltip is the red box explaining why

**Comment by lwatson2016** [comment]
Hi, my team would like to work on this issue.

**Comment by duncanspumpkin** [comment]
@lwatson2016 this may require looking at the vanilla game to confirm changes. Let me know if you need help.

**Comment by joshtucker132** [comment]
I'm working with @lwatson2016 on this issue. We think that the problem occurs when evaluating whether the small scenery object is being placed on water. The isOnWater bool never evaluates to true and therefore the correct tooltip is not displayed. 

What should the base height and target height be for an underwater tile?

**Comment by ZehMatt** [comment]
I think I did look into this before, it did seem as the order of checks have been changed from the original which results in the different error message.

**Comment by joshtucker132** [comment]
Which is the criteria for placing objects on water tiles? Objects like standard trees and hedges give an error message, but small trees and bushes can be placed on top of the water.

**Comment by duncanspumpkin** [comment]
Not too sure but look at the small scenery place action. After the call to map can construct with clear there should be the code for the error messages.

**Comment by duncanspumpkin** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/actions/SmallSceneryPlaceAction.hpp#L155

Issue will be in this section.

**Comment by duncanspumpkin** [comment]
Trees can't be stacked. I expect that is the flag you will need to look at.

**Comment by joshtucker132** [comment]
I believe the issue is that the bool isOnWater which determines whether "level land required" (isOnWater=false) or "can only build this on land" (isOnWater=true) evaluates incorrectly. By changing L136 to != the "can only build this on land" is displayed when placing the tree on water.

https://github.com/OpenRCT2/OpenRCT2/blob/25eb403c78c7a90be6ba3b63cc821660bb268178/src/openrct2/actions/SmallSceneryPlaceAction.hpp#L136-L139

**Comment by duncanspumpkin** [comment]
_loc.z is 0 when first trying to place an item. And means it will try place on the surface or surface of the water. It is non zero when shift or ctrl is pressed.

**Comment by joshtucker132** [comment]
When I put in debug statements, _loc.z is 96 and the isOnWater bool does not evaluate to true when it should.

**Comment by duncanspumpkin** [comment]
Ahh the issue must be with the caller `sub_6E1F34`  L1431 of TopToolbar.cpp. See how it adjusts the height if on water.

**Comment by duncanspumpkin** [comment]
`0xF64F1D` is `isOnWater`. This is taken from the original for lines https://github.com/OpenRCT2/OpenRCT2/blob/25eb403c78c7a90be6ba3b63cc821660bb268178/src/openrct2/actions/SmallSceneryPlaceAction.hpp#L187-L204

```
CPU Disasm
Address   Hex dump          Command                                  Comments
006E0B3F  |. /75 37         JNZ SHORT 006E0B78
006E0B41  |. |66:833D CA4EF CMP WORD PTR DS:[0F64ECA],0
006E0B49  |. |74 2D         JE SHORT 006E0B78
006E0B4B  |. |803D 1D4FF600 CMP BYTE PTR DS:[0F64F1D],0
006E0B52  |. |0F85 DB010000 JNE 006E0D33 // 0x40A Error aka STR_CAN_ONLY_BUILD_THIS_ON_LAND
006E0B58  |. |F646 04 1F    TEST BYTE PTR DS:[ESI+4],1F
006E0B5C  |. |0F85 DC010000 JNZ 006E0D3E // 0x530 Error aka STR_LEVEL_LAND_REQUIRED
006E0B62  |. |66:0FB66E 02  MOVZX BP,BYTE PTR DS:[ESI+2]
006E0B67  |. |66:C1E5 03    SHL BP,3
006E0B6B  |. |66:3B2D C84EF CMP BP,WORD PTR DS:[0F64EC8]
006E0B72  |. |0F85 C6010000 JNE 006E0D3E
006E0B78  |> \53            PUSH EBX
```

**Comment by joshtucker132** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/cc90c3b0e115d595601e6f48701759406218a4ff/src/openrct2-ui/windows/TopToolbar.cpp#L1426-L1431

Based on this code, it seems that _loc.z would not be 0 unless the water height is 0, which I would guess is never true? So this appears to be where the issue is since _loc.z needs to be 0 in order for isOnWater to evaluate correctly.

**Comment by duncanspumpkin** [comment]
Yes I had the same thought and then checked it against vanilla. The problem has to be in game action. I am pretty confident in my above guess that it is L190 that should change.

**Comment by joshtucker132** [comment]
So you are saying L190 should be if(isOnWater)?

**Comment by duncanspumpkin** [comment]
I think so. Have a play about and see if that works.

**Comment by joshtucker132** [comment]
I have tested that before and it does work, but in my opinion it is not solving the root of the issue. If `isOnWater = true` the most logical tooltip would be *can only build this on land* but with your fix that condition would evaluate to *level land required*. It still seems to me that the best fix would be to figure out how to have `isOnWater` evaluate correctly.

**Comment by duncanspumpkin** [comment]
Well to fix that you will need to adjust `sub_6E1F34` L1431 so that it doesn't set the z value to a value that is non zero.

**Comment by joshtucker132** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/25eb403c78c7a90be6ba3b63cc821660bb268178/src/openrct2/actions/SmallSceneryPlaceAction.hpp#L132-L140

Is this code redundant? It seems that `if (baseHeight & 0xFFFF0000)` is a valid check for determining whether the object is on water, so including `if (_loc.z == 0)` is unnecessary.

**Comment by lwatson2016** [comment]
What is the definitions of waterheight, targetheight and baseheight?
I imagine waterheight is the height between floor of the lake and water surface, is this correct? What do these other two variables mean?
Why is the 'Cannot build underwater' error message returned when waterheight > targetheight at L158?

https://github.com/OpenRCT2/OpenRCT2/blob/9554918e30c7366e6c00329ba12f206b7cb0ffe3/src/openrct2/actions/SmallSceneryPlaceAction.hpp#L155-L161

**Comment by ZehMatt** [comment]
I believe targetHeight is where its supposed to be placed. So if targetHeight < waterHeight or waterHeight > targetHeight it means its actually underwater.

**Comment by duncanspumpkin** [comment]
This game action has two modes. One where a height is specified and one where it is not. When not specified loc.z is zero and the height will be the surface or water surface. So the is on water bool is not just telling you if it's on water it is telling you it is on the water and we are in place on surface mode. This makes it awkward to change without making knock on mistakes.

**Comment by duncanspumpkin** [comment]
Target height is the loc.z height in normal mode and it is the surface or water surface height when in place on surface mode. 

It represents the height the item will be placed at. 

Base height is badly named it is the return of get height. And represents both the surface height and the water height. Stuffed into the same variable shifted by 16.

**Comment by joshtucker132** [comment]
Is it correct to say that loc.z and base height should be the same if on water?

**Comment by lwatson2016** [comment]
How can we know which mode the game action is in? Is there some variable or file you can direct us to?

This seriously affects isOnWater boolean because it is evaluated by a comparison of loc.z.

**Comment by duncanspumpkin** [comment]
Top toolbar is the function that calls this. I don't actually know why it's setting the value to non zero. It doesn't make sense. 

The top toolbar code is quite hard to follow. It branches in a couple different ways depending on if ctrl or shift is pressed.

**Comment by joshtucker132** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/cc90c3b0e115d595601e6f48701759406218a4ff/src/openrct2-ui/windows/TopToolbar.cpp#L1426-L1431

These lines appear to be where the problem is occurring. Are they correct in the logic? Is `GetWaterHeight()` supposed to return a zero value (currently returns non-zero) if neither shift or ctrl is pressed? If L1427 is correct, then would the easiest fix be to remove L1428-1431 to prevent `gSceneryPlaceZ` from being updated to a non-zero value?

**Comment by duncanspumpkin** [comment]
I'm pretty confident that it should be zero if neither ctrl and or shift are pressed. Have a play and see how it looks

**Comment by joshtucker132** [comment]
We have tested this already and by removing L1428-1431, `gSceneryPlaceZ` will always be zero (when shift or ctrl is not pressed) causing the correct tooltip to display for all trees. Is this the best fix?

**Comment by duncanspumpkin** [comment]
I think it makes sense. It would simplify the code. Can you pr it with those lines removed
